### PR TITLE
Cleanup PrefixTrim

### PR DESCRIPTION
**Pull request by lixinchengdu**
## Overview

Cleanup PrefixTrim related code. PrefixTrim is not compatible with the new GC mechanism.

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java [hunk 1]
```diff
@@ -30,15 +30,13 @@
 public class InMemoryStreamLog implements StreamLog {
 
     private Map<Long, LogData> logCache;
-    private volatile long startingAddress;
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java]: still in the case of sparse trim the starting address can change, isn't this relevant to keep track of?

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java [hunk 2]
```diff
@@ -69,12 +69,9 @@ public static boolean handleFailure(@Nonnull IReconfigurationHandlerPolicy failu
      */
     private static Duration getStateTransferTimeoutEstimate(CorfuRuntime runtime) {
 
-        // Try to estimate a reasonable timeout to rebuild the logging unit
-        Token trimMark = runtime.getAddressSpaceView().getTrimMark();
-
         long tail = runtime.getAddressSpaceView().getLogTail();
 
-        long rangeToReplicate = tail - trimMark.getSequence();
+        long rangeToReplicate = tail - 0L;
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java]: -0L? we can remove this.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java (hunk 1 above)
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java]: I don't know why Wenbin's change is in this PR. @WenbinZhu Could you help answer this question?

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java (hunk 2 above)
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java]: I don't know why Wenbin's change is in this PR. @WenbinZhu Could you help answer this question?

Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java]: This is not my change.

File: infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java (hunk 1 above)
Comment (WenbinZhu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java]: This is not my change.

Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/log/InMemoryStreamLog.java]: I see why I delete this line. startingAddress is used for prefixTrim for streamLog. For the new GC, log data (as well as garbage data) are trimmed at the logunit's own pace. As as result, startingAddress is not needed for externally triggering compaction.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java (hunk 2 above)
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java]: I have deleted 0L for conciseness.

File: infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java [hunk 3]
```diff
@@ -237,11 +234,6 @@ private TxResolutionResponse txnCanCommit(TxResolutionInfo txInfo) {
             return new TxResolutionResponse(TokenType.TX_ABORT_NEWSEQ);
         }
 
-        if (txSnapshotTimestamp.getSequence() < trimMark) {
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServer.java]: I checked my recent patch and there isn't  any check like this in the recent patch on sequencer trim. In the previous implementation, the prefixTrim would invalidate transaction resolution information before trimMark and that's is why this check makes sense here. In my recent patch, I deleted the prefix trim of transaction resolution information. The most most recent implementation relies on the some other checks to ensure correctness. Anyway, transaction resolution information is stored in cache, we must have other mechanism other than simply checking trimMark to ensure correctness.

File: infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java]: I agree! The more actual state transfer time relies on how the new implementation is implemented. I will add a TODO in the code to remind the developer about this problem.

File: infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java [hunk 4]
```diff
@@ -72,19 +72,8 @@ public static void transfer(Layout layout,
...
-                                    + "greater than end of segment {}",
-                            trimMark, segment.getEnd());
-                    return true;
-                }
-
-                // State transfer should start from segment start address or trim mark
-                // whichever is higher.
-                final long segmentStart = Math.max(trimMark, segment.getStart());
+                // State transfer should start from segment start address
+                final long segmentStart = segment.getStart();
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java]: Same here. I will add a TODO here.

File: runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java [hunk 5]
```diff
@@ -531,7 +528,7 @@ private void applyForEachAddress(BiConsumer<Long, ILogData> logDataProcessor) {
 
                 invokeNecromancer(range, logDataProcessor);
 
-            } catch (TrimmedException ex) {
+            } catch (Exception ex) {
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/recovery/FastObjectLoader.java]: I reverted it back to TrimmedException.

**Pull request merged**